# MAGIC
To considerable heterogeneity within each cell type, we introduce the single-cell polygenic burden score (scPBS) in our MAGIC toolkits, a method to evaluate polygenic burden enrichment of rare variants in individual cells of scRNA-seq data. scPBS assesses whether a given cell has an excess of rare PTVs among EM cases in cell-specific highly expressed genes derived from scRNA-seq using an appropriately matched empirical null distribution to estimate well-calibrated P values. scPBS consists of three steps. First, scPBS constructs a set of cell-specific highly expressed genes, the most specific genes in each individual cell from high cell-to-cell variable features in the dataset. Second, scPBS selects qualifying variants (QV) that pass all filters of a specific model and builds the geneset(cell)-by-individual indicator matrix used for the polygenic burden test. Third, scPBS quantifies the EM-related polygenic burden coefficient (odds ratio) of highly expressed genes in each cell to generate single-cell specific polygenic burden scores (scPBS). Following previous studies, we correlate scPBS with the expression level of each gene across cells and prioritize the trait-relevant genes by ranking the Pearson correlation coefficients (PCCs). Finally, a rare variant trait-relevant score (rvTRS) of each cell is computed by averaging the expression level of the top 10% trait-relevant genes based on ranked PCCs and subtracting the random control cell score via the cell-scoring method used in Seurat.
